include/deal.II/lac/trilinos_vector.h

```diff
@@ -931,11 +931,93 @@ namespace TrilinosWrappers
 
 #endif
 
+} /* namespace TrilinosWrappers */
 
-}
+/*@}*/
 
 
-/*@}*/
+namespace internal
+{
+  template <typename> class ReinitRangeFactory;
+  template <typename> class ReinitDomainFactory;
+
+  /*
+   * A factory class internally used in linear_operator.h.
+   * Specialization for TrilinosWrappers::MPI::Vector.
+   */
+  template<>
+  class ReinitRangeFactory<TrilinosWrappers::MPI::Vector>
+  {
+  public:
+    template <typename Matrix>
+    std::function<void(TrilinosWrappers::MPI::Vector &, bool)>
+    operator()(const Matrix &matrix)
+    {
+      return [&matrix](TrilinosWrappers::MPI::Vector &v, bool fast)
+      {
+        v.reinit(matrix.range_partitioner(), fast);
+      };
```

Comment:
I think I don't understand why you wrap these functions into a structure. Why  not have  functions `reinit (matrix,vector,fast)` for different vector types and general matrix template arguments, and then just create the function object in `linop` using `std::bind (&internal::reinit, matrix, _1, _2)`?